### DISCUSSION TITLE
Accessing Google APIs with a Google service account

### DISCUSSION THREAD

**Discussion by jmgilman** [Budibase/budibase#4218] [Help] [answered]
Hi,

I'm currently evaluating Budibase for internal usage but have run into a seemingly difficult roadblock: a large portion of our data is coming from Google Cloud Functions with an API gateway in front of them. The API gateway is secured using IAM and external requests must include a signed JWT token in order to access the endpoints.

The only way to access the API endpoints externally is to use a private key associated with a service account to sign a JWT and include it with the external HTTP request. This seems a bit too complex for what Budibase can handle, but before assuming it's not possible I wanted to reach out to see if someone has integrated with Google Cloud Functions already and can detail their setup.

**Comment by shogunpurple** [comment]
Hi @jmgilman - thank you for typing up this detailed issue. We just recently shipped our new and improved REST connector, but sadly it does not account for this kind of auth that would allow you to sign a JWT from within the platform. Just to clarify - essentially what you would be looking for is for budibase to programmatically call something like this (the CLI command), except obviously we would use the SDK instead:
`curl  -H "Authorization: bearer $(gcloud auth print-identity-token)"  https://REGION-PROJECT_ID.cloudfunctions.net/FUNCTION_NAME`

On return of this token, it would inject that into the calls to the cloud functions and allow authenticated access?

**Comment by jmgilman** [comment, marked as answer]
Hi @shogunpurple,

It would probably be in the best interest of everyone to make this support as generic as possible since this type of authentication is not uncommon. For example, [Bubble provides this type of authentication support](https://manual.bubble.io/account-and-marketplace/building-plugins/adding-api-connections#json-web-token).

While Google provides an SDK for automating signing the JWT, it's possible to manually implement the signing since there's nothing proprietary about it. You'd essentially just need to collect enough information from the end-user about how to create the JWT (i.e. private key, scope, audience, etc.) and then add code to handle generating a JWT for every request to the respective API endpoint.

**Reply by shogunpurple** [reply]
Hi @jmgilman - this makes total sense, and thank you for the relevant information. I'm going to convert this into an issue as it's tangible work - we will try to slot it in in the future roadmap, but I can't give you a timeline unfortunately. Thank you!

**Reply by jmgilman** [reply]
Thanks!

**Reply by mielune** [reply]
I have the same need in #5951 based on an internal OIDC provider.

BB got a jwt access token then throw its subscription to some apis, will allow bb apps to get access to some of this APIs.

Have you the issue number ?

Thanks

**Reply by mielune** [reply]
Hi @shogunpurple,

Did you created an issue related to this feature ?

I did a schema for collegues about this process, I share it with you here :)

![bb-jwt](https://user-images.githubusercontent.com/8350742/171831088-29005404-a958-4d62-9ffe-b66bc7d80ccd.png)

**Reply by shogunpurple** [reply]
Hi @mielune - we have! It will be in the next (June) release.

https://github.com/Budibase/budibase/issues/3827

**Reply by einarpersson** [reply]
I am currently evaluating Budibase. I am also using a Service Account in my google organization (i.e. not the "current user"), but it sounds like supporting something like that is not part of the June release? (@shogunpurple )

What I hoped for was that there would be some kind of generic JWT-adapter, or even being able to access crypto functions directly (I know n8n exposes this which makes it very flexible).

I hope Budibase keeps evolving!
